### PR TITLE
Introduce RPC for roles

### LINKED ISSUE #9842: Introduce RPC for roles
Currently, it is quite problematic to execute a function provided by a role on one instance from another instance. We need to provide a way to do this.

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch-set introduces RPC for roles from the same config.

Closes #9842

### CODE REVIEW

File: src/box/lua/experimental/connpool.lua [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        local conn = connect_using_config(instance_name, candidate_opts)
+        if conn:wait_connected() then
+            return conn
+        end
+    end
+    error('Connection to candidate instances failed', 0)
+end
+
+local function call_role(role_name, func_name, args, opts)
+    checks('string', 'string', '?table', {
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua:143]: I was planning to add them a little later as I find it a little weird to add them directly to the opts. But I have no problem adding another patch for them here.

File: src/box/lua/config/init.lua [hunk 2]
```diff
@@ -578,6 +578,34 @@ function methods.instances(self)
...
+    end
+    if opts.show_password ~= nil and type(opts.show_password) ~= 'boolean' then
+        error(('Expected boolean, got %s'):format(opts.show_password), 0)
+    end
+
+    local uri_opts = {
+        instance = opts.instance,
+        use_default = true,
+    }
+    local uri = table.copy(self._configdata:_instance_uri(uri_type, uri_opts))
```
Comment (Totktonada) [inline src/box/lua/config/init.lua:602]: Shouldn't it be `_configdata_applied` for consistency with `config:get()`?

File: changelogs/unreleased/gh-9842-rpc-for-roles.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+## feature/config
+
+* Introduced `instance_uri()` method for `config` module (gh-9842).
+* Introduced module `experimental.connpool` and function
+  `connect_using_config()` (gh-9842).
+* Introduced function `get_candidates()` from `experimental.connpool`
+  module (gh-9842).
+* Introduced function `call_role()` from `experimental.connpool`
+  module (gh-9842).
```
Comment (andreyaksenov) [inline changelogs/unreleased/gh-9842-rpc-for-roles.md:9]:
```suggestion
* Introduced the `instance_uri()` method for the `config` module (gh-9842).
* Introduced the `experimental.connpool` module and the
  `connect_using_config()` function (gh-9842).
* Introduced the `get_candidates()` function in the `experimental.connpool`
  module (gh-9842).
* Introduced the `call_role()` function in the `experimental.connpool`
  module (gh-9842).
```

**Review by andreyaksenov — APPROVED** [review]

File: src/box/lua/init.c [hunk 4]
```diff
@@ -169,15 +169,15 @@ extern char session_lua[],
...
 	config_source_etcd_lua[],
 	config_storage_init_lua[],
 	config_source_storage_lua[],
-	config_extras_lua[]
+	config_extras_lua[],
 #endif
-	;
+	experimental_connpool_lua[];
+
 	/* }}} config */
```
Comment (Totktonada) [inline src/box/lua/init.c:181]: Nit: This new item should be outside of the `/* {{{ config */` `/* }}} config */` block.

File: src/box/lua/experimental/connpool.lua [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+local function connect_using_config(instance_name, opts)
+    checks('string', {
+        wait_connected = '?boolean|number',
+        fetch_schema = '?boolean',
+    })
+    opts = opts or {}
+
+    local conn = connections[instance_name]
+    if conn == nil or conn.state == 'error' or conn.state == 'closed' or
+       conn.space == nil and opts.fetch_schema then
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:16]: I'm not sure about this check. First, `opts.fetch_schema` may be `nil` (and it is assumed as `true`). So, it should be `opts.fetch_schema ~= false` at least.

Second, it is possible that a space information is in fly. I guess that it is possible that space information fetching may be slower than retrying of the calls to `connect_using_config`. It means that a new connection will be created each time.

File: src/box/lua/experimental/connpool.lua [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    -- wait_connected == true is not supported.
+    local wait_connected =
+        opts.wait_connected ~= true and opts.wait_connected or nil
+
+    assert(conn ~= nil)
+    local ok, res = pcall(conn.wait_connected, conn, wait_connected)
+    if not ok then
+        local msg = 'Unable to connect to instance %q: %s'
+        error(msg:format(instance_name, res.message), 0)
+    end
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:52]: First, I think that if something is not supported, it should lead to an error.

Next, it is a bit hard to understand that the `wait_connected` variable contains a timeout for creating a connection. I would check the type and assign it to a variable named as `connect_timeout`.

BTW, the API may be more straightforward if we add it as a separate `connect_timeout` option into the API. It is completely up to you, just my though.

File: changelogs/unreleased/gh-9842-rpc-for-roles.md [hunk 7]
```diff
@@ -1,3 +1,5 @@
 ## feature/config
 
 * Introduced `instance_uri()` method for `config` module (gh-9842).
+* Introduced module `experimental.connpool` and function
+  `connect_using_config()` (gh-9842).
```
Comment (Totktonada) [inline changelogs/unreleased/gh-9842-rpc-for-roles.md:5]: The `config:` prefix in commit messages regarding the connection pool module seems irrelevant. The same for the `## feature/config` changelog category.

File: changelogs/unreleased/gh-9842-rpc-for-roles.md [hunk 8]
```diff
@@ -1,3 +1,5 @@
 ## feature/config
```
Comment (Totktonada) [inline changelogs/unreleased/gh-9842-rpc-for-roles.md:1]: > NO_DOC=experimental feature

We agreed to have some brief documenation for experimental features.

File: src/box/lua/experimental/connpool.lua [hunk 9]
```diff
@@ -54,6 +54,45 @@ local function connect_using_config(instance_name, opts)
...
+    for _, label in pairs(labels) do
+        if instance_labels[label] == nil then
+            return false
+        end
+    end
+    return true
+end
+
+local function get_candidates(role_name, opts)
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:72]: In a voice discussion with Mergen we agreed that filtering by a role should be one of the filters, not some kind of a primary filter in a separate argument.

So, we likely will be somethings like the following.

```lua
pool.get_candidates({
    role = 'myrole',
    labels = {zone = 'msk'},
})
```

Or the following.

```lua
pool.get_candidates({
    roles = {'myrole1', 'myrole2'},
    labels = {zone = 'msk'},
})
```

File: src/box/lua/experimental/connpool.lua [hunk 10]
```diff
@@ -54,6 +54,45 @@ local function connect_using_config(instance_name, opts)
...
+        return true
+    end
+    if instance_labels == nil then
+        return false
+    end
+    for _, label in pairs(labels) do
+        if instance_labels[label] == nil then
+            return false
+        end
+    end
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:68]: As I see, we check a presence of the label, but we don't verify that it has a value given by a caller.

It is easy to imagine that one have the `zone` label and wants to filter instances from a specific zone.

I propose to implement the `labels` filter with the following semantic.

The instance `I` is accepted by the `opts.labels` filter if `I.labels` is *covered* by `opts.labels`. It means that each key-value pair from `opts.labels` has the same key-value pair in `I.labels`.

If we will need a filter that verifies a presence of a label without taking into account its value, I would add it later with a name like `has_labels`.

```lua
pool.get_candidates({
    has_labels = {'mylabel1', 'mylabel2'},
})
```

File: src/box/lua/experimental/connpool.lua [hunk 11]
```diff
@@ -54,6 +54,45 @@ local function connect_using_config(instance_name, opts)
...
+        local roles = iconfig.roles or {}
+        if is_labels_match(opts.labels, iconfig.labels) then
+            for _, role in pairs(roles) do
+                if role == role_name then
+                    table.insert(candidates, instance_name)
+                    break
+                end
+            end
+        end
+    end
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:91]: I would structurize the code in a way that allows to decouple the filters. For example,

```lua
local function filter_by_role(instance_name)
    if opts.role == nil then
        return true
    end
    local roles = config:get('roles', {instance = instance_name})
    return table_has_key(roles, opts.role)
end

local function filter_by_labels(instance_name)
    if opts.labels == nil then
        return true
    end
    local labels = config:get('labels', {instance = instance_name})
    return table_is_covered_by(labels, opts.labels)
end

return fun.iter(config:instances())
    :filter(filter_by_role)
    :filter(filter_by_labels)
    :totable()
```

(Extracting the `table_{has/is}_*` helpers would also increase readability, IMHO.)

(I don't test this code, it is just illustration.)

I don't think that repeated `config:get()` call is a problem. If it is, it is better to implement some caching inside the call.

File: src/box/lua/experimental/connpool.lua [hunk 12]
```diff
@@ -92,7 +92,70 @@ local function get_candidates(role_name, opts)
...
+        end
+        local msg = 'No instances with role %q are available'
+        error(msg:format(role_name), 0)
+    end
+
+    if opts.prefer_local == nil or opts.prefer_local == true then
+        local name = box.info.name
+        for n, candidate_name in pairs(candidates) do
+            if candidate_name == name then
+                table.remove(candidates, n)
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:119]: We shoudn't modify a table during `pairs()`.

File: src/box/lua/experimental/connpool.lua [hunk 13]
```diff
@@ -92,7 +92,70 @@ local function get_candidates(role_name, opts)
...
+    end
+
+    while #candidates > 0 do
+        local n = math.random(#candidates)
+        local instance_name = table.remove(candidates, n)
+        local candidate_opts = {wait_connected = false}
+        local conn = connect_using_config(instance_name, candidate_opts)
+        if conn:wait_connected() then
+            return conn
+        end
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:137]: I would let it fail if it doesn't became connected within the given/default connect timeout. It sounds better than an infinite waiting.

File: src/box/lua/experimental/connpool.lua (hunk 12 above)
Comment (Totktonada) [reply src/box/lua/experimental/connpool.lua:119]: BTW, this code would be simpler if we just check `candidates` using a helper like `table_has_key` (that I proposed in comments for one of the previous patches).

File: src/box/lua/experimental/connpool.lua (hunk 13 above)
Comment (Totktonada) [reply src/box/lua/experimental/connpool.lua:137]: Or `conn:wait_connected()` is not infinite?

BTW, if it is infinite, there is no point in the `while` loop here.

File: src/box/lua/experimental/connpool.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: @olegrok I added these options.

File: changelogs/unreleased/gh-9842-rpc-for-roles.md (hunk 3 above)
Comment (ImeevMA) [reply changelogs/unreleased/gh-9842-rpc-for-roles.md]: Thanks, fixed!

File: src/box/lua/init.c (hunk 4 above)
Comment (ImeevMA) [reply src/box/lua/init.c]: Fixed.

File: src/box/lua/experimental/connpool.lua (hunk 5 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: Changed the condition.

File: src/box/lua/experimental/connpool.lua (hunk 6 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: After discussion we decided to accept only boolean for `wait_connected` and add another numeric option `connect_timeout`. These options are passed directly to `net.box.connect()`. These two options works about the same way as `wait_connected` option for `cartridge.pool.connect()`. There is couple of cases when the behavior is different:
1) If `wait_connected` option for `cartridge.pool.connect()` more than default `connect_timeout` of `net.box.connect()`, than default `connect_timeout` is actually used;
2) If `wait_connected` option for `cartridge.pool.connect()` less than default `connect_timeout` of `net.box.connect()`, it was actually possible that an error returned before the connection is successfully established. This actually seems to be expected behavior, as the error says `Connection not established (yet)`.

File: changelogs/unreleased/gh-9842-rpc-for-roles.md (hunk 7 above)
Comment (ImeevMA) [reply changelogs/unreleased/gh-9842-rpc-for-roles.md]: Changed to `experimental`.

File: src/box/lua/experimental/connpool.lua (hunk 11 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: Reworked this part.

File: src/box/lua/experimental/connpool.lua [hunk 14]
```diff
@@ -92,7 +92,70 @@ local function get_candidates(role_name, opts)
...
+        prefer_local = '?boolean',
+    })
+    opts = opts or {}
+
+    local conn = get_connection(role_name, {
+        prefer_local = opts.prefer_local,
+        labels = opts.labels,
+    })
+    assert(conn ~= nil)
+    return conn:call(('%s.%s'):format(role_name, func_name), args)
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: Renamed the function to `call` and moved role to options.

File: src/box/lua/experimental/connpool.lua (hunk 12 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: True, fixed. For now I did not implemented `table_has_key()`.

File: src/box/lua/experimental/connpool.lua [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        wait_connected = '?boolean',
+        fetch_schema = '?boolean',
+    })
+    opts = opts or {}
+
+    local conn = connections[instance_name]
+    if not is_connection_valid(conn, opts) then
+        local uri_opts = {
+            instance = instance_name,
+            show_password = true,
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:31]: AFAIR, this `:instance_uri()` option is gone (was present in a draft patch version).

File: src/box/lua/experimental/connpool.lua [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    -- established or an error occurs (including a timeout error).
+    if opts.wait_connected ~= false and conn:wait_connected() == false then
+        local msg = 'Unable to connect to instance %q: connection timeout'
+        error(msg:format(instance_name), 0)
+    end
+    return conn
+end
+
+return {
+    connect_using_config = connect_using_config,
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:194]: It appears that all the functions in the module takes information from config. So the `_using_config` suffix now looks, hm, inconsistent. It should be either in all the functions or neither.

I propose to drop it and name the function `connect`.

And, while we're here, if we do something with `get_candidates`, all the functions will be one-word-named, which would look nice. For example, `connect`, `filter` and `call`.

Just thoughts. Feel free to ignore.

**Review by Totktonada — APPROVED** [review]
It looks OK for me at whole. Thank you for working on it!

There are some thoughts around. Please, consider comments above.

File: src/box/lua/experimental/connpool.lua [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+local function is_connection_valid(conn, opts)
+    if conn == nil or conn.state == 'error' or conn.state == 'closed' then
+        return false
+    end
+    assert(type(opts) == 'table')
+    local conn_opts = conn.opts or {}
+    if opts.fetch_schema == true and conn_opts.fetch_schema == false then
+        return false
+    end
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: Thanks, fixed.

File: src/box/lua/experimental/connpool.lua (hunk 15 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: True, fixed.

File: src/box/lua/experimental/connpool.lua [hunk 18]
```diff
@@ -116,7 +116,82 @@ local function get_candidates(opts)
...
+        local candidate_opts = {wait_connected = false}
+        local conn = connect_using_config(instance_name, candidate_opts)
+        if conn:wait_connected() then
+            return conn
+        end
+    end
+    return nil
+end
+
+local function call(func_name, args, opts)
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua:145]: Added.

File: src/box/lua/experimental/connpool.lua (hunk 16 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua]: Fixed.

File: src/box/lua/experimental/connpool.lua [hunk 19]
```diff
@@ -0,0 +1,191 @@
+local config = require('config')
+local checks = require('checks')
+local netbox = require('net.box')
+
+local connections = {}
```
Comment (olegrok) [inline src/box/lua/experimental/connpool.lua:5]: Is there any mechanism to clean this table? It is possible to overflow it somehow during long runtime?

File: src/box/lua/experimental/connpool.lua [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+    local candidates = table.copy(all_candidates)
+    if prefer_local ~= false then
+        local candidate_idx = nil
+        for n, candidate in ipairs(candidates) do
+            if candidate == box.info.name then
+                candidate_idx = n
+                local conn = connect(box.info.name, {wait_connected = false})
+                if conn:wait_connected() then
+                    return conn
+                end
```
Comment (olegrok) [inline src/box/lua/experimental/connpool.lua:125]: Is not it the same as just connect with wait_connected = true?

Comment (Totktonada) [reply src/box/lua/experimental/connpool.lua:125]: I had the same question, but then found that if we'll do `connect()` without options, we need to pcall it to try the next candidate.

File: src/box/lua/experimental/connpool.lua (hunk 19 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua:5]: At the moment the only way to clear the table is to restart Tarantool. As for overflow, we have standard Lua problems here. What I mean is that if you're using a GC64, you have a maximum of 2^26 keys, otherwise you'll likely hit the 2GB memory limit sooner.

File: src/box/lua/experimental/connpool.lua (hunk 20 above)
Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua:125]: Also, `connect()` returns an already existing connection, if it exists.

File: src/box/lua/experimental/connpool.lua (hunk 19 above)
Comment (olegrok) [reply src/box/lua/experimental/connpool.lua:5]: The main problem here that you can silently get overflow error and it will be very hard to understand exact reason.

**Review by olegrok — APPROVED** [review]

File: src/box/lua/experimental/connpool.lua (hunk 19 above)
Comment (Totktonada) [reply src/box/lua/experimental/connpool.lua:5]: I don't think that we can realistically reach any Lua memory/Lua table size limit here. However, I agree that it would be good to delete items that are not in the config anymore. We can do it in `connect()`.

I wouldn't insist on it in the scope of this PR.

File: src/box/lua/experimental/connpool.lua [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+local function connect(instance_name, opts)
+    checks('string', {
+        connect_timeout = '?number',
+        wait_connected = '?boolean',
+        fetch_schema = '?boolean',
+    })
+    opts = opts or {}
+
+    local conn = connections[instance_name]
```
Comment (Totktonada) [inline src/box/lua/experimental/connpool.lua:27]: Yet another thing that I would solve in a separate PR.

It seems, if instance's URI is changed, we don't re-create a connection.

Comment (ImeevMA) [reply src/box/lua/experimental/connpool.lua:27]: I created a new issue about this: https://github.com/tarantool/tarantool/issues/9910
